File: src/main/java/org/flossware/civilization/module/ReligionModule.java
package org.flossware.civilization.module;

import org.flossware.civilization.model.ReligionState;
import org.flossware.civilization.model.Event;
import org.flossware.civilization.model.Event.EventSeverity;
import org.flossware.civilization.model.Event.EventType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.SplittableRandom;

/**
 * Religion and cultural dynamics module.
 *
 * Pure function: (state, tradeConnectivity, random) → (newState, events)
 */
public final class ReligionModule {

    private static final double SCHISM_THRESHOLD = 0.05;
    private static final double MINORITY_THRESHOLD = 0.2;
    private static final double UNITY_THRESHOLD = 0.6;

    /**
     * Ticks religion forward by one time step.
     *
     * @param current Current religion state
     * @param tradeConnectivity Trade connectivity factor (0.0 to 1.0+)
     * @param random Module-specific random generator
     * @return Updated state and events
     */
    public static ModuleResult<ReligionState> tick(
        ReligionState current,
        double tradeConnectivity,
        SplittableRandom random
    ) {
        List<Event> events = new ArrayList<>();
        int year = 0; // Will be set by engine

        // Spread religions based on spreadRate and trade
        Map<String, Double> newShares = spreadReligions(
            current.religionShares(),
            current.spreadRate(),
            tradeConnectivity,
            random
        );

        // Calculate unity (largest religion share)
        double unity = newShares.values().stream()
            .mapToDouble(d -> d)
            .max()
            .orElse(0.0);

        // Calculate stability bonus
        double stabilityBonus = unity * 0.2;

        // Check for schism
        boolean schismOccurred = checkForSchism(newShares, unity, random);
        if (schismOccurred) {
            events.add(new Event(
                year,
                "",
                EventType.RELIGIOUS_SCHISM,
                EventSeverity.MAJOR,
                "Religious schism splits the dominant faith!",
                null
            ));

            // Split the largest religion
            newShares = splitLargestReligion(newShares, random);

            // Recalculate unity after schism
            unity = newShares.values().stream()
                .mapToDouble(d -> d)
                .max()
                .orElse(0.0);

            // Recalculate stability bonus
            stabilityBonus = unity * 0.2;
        }

        ReligionState newState = new ReligionState(
            newShares,
            unity,
            stabilityBonus,
            current.spreadRate()
        );

        return new ModuleResult<>(newState, events);
    }

    /**
     * Spread religions based on spread rate and trade connectivity.
     * Religions with higher shares grow slightly, influenced by trade.
     */
    private static Map<String, Double> spreadReligions(
        Map<String, Double> currentShares,
        double spreadRate,
        double tradeConnectivity,
        SplittableRandom random
    ) {
        Map<String, Double> newShares = new HashMap<>(currentShares);

        // Trade increases religious exchange
        double effectiveSpreadRate = spreadRate * (1.0 + tradeConnectivity * 0.5);

        // Apply small random fluctuations to each religion
        for (Map.Entry<String, Double> entry : currentShares.entrySet()) {
            String religion = entry.getKey();
            double currentShare = entry.getValue();

            // Random fluctuation: -spreadRate to +spreadRate
            double change = (random.nextDouble() * 2.0 - 1.0) * effectiveSpreadRate;
            double newShare = Math.max(0.0, Math.min(1.0, currentShare + change));

            newShares.put(religion, newShare);
        }

        // Normalize shares to sum to 1.0
        return normalizeShares(newShares);
    }

    /**
     * Check if a religious schism occurs.
     * Conditions: minority religion > 20%, unity < 60%, and 5% random chance.
     */
    private static boolean checkForSchism(
        Map<String, Double> shares,
        double unity,
        SplittableRandom random
    ) {
        // Check if any minority religion exceeds threshold
        boolean hasSignificantMinority = shares.values().stream()
            .anyMatch(share -> share > MINORITY_THRESHOLD && share < unity);

        // Schism occurs if conditions are met and random check passes
        return hasSignificantMinority
            && unity < UNITY_THRESHOLD
            && random.nextDouble() < SCHISM_THRESHOLD;
    }

    /**
     * Split the largest religion into two factions.
     */
    private static Map<String, Double> splitLargestReligion(
        Map<String, Double> shares,
        SplittableRandom random
    ) {
        Map<String, Double> newShares = new HashMap<>(shares);

        // Find the largest religion
        String largestReligion = shares.entrySet().stream()
            .max(Map.Entry.comparingByValue())
            .map(Map.Entry::getKey)
            .orElse(null);

        if (largestReligion != null) {
            double originalShare = shares.get(largestReligion);

            // Split roughly 60/40 with some randomness
            double splitRatio = 0.6 + (random.nextDouble() * 0.2 - 0.1); // 0.5 to 0.7
            double mainShare = originalShare * splitRatio;
            double schismShare = originalShare * (1.0 - splitRatio);

            newShares.put(largestReligion, mainShare);
            newShares.put(largestReligion + " (Reformed)", schismShare);
        }

        return normalizeShares(newShares);
    }

    /**
     * Normalize religion shares to sum to 1.0.
     */
    private static Map<String, Double> normalizeShares(Map<String, Double> shares) {
        double total = shares.values().stream().mapToDouble(d -> d).sum();

        if (total <= 0) {
            return shares;
        }

        Map<String, Double> normalized = new HashMap<>();
        for (Map.Entry<String, Double> entry : shares.entrySet()) {
            normalized.put(entry.getKey(), entry.getValue() / total);
        }

        return normalized;
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/module/ReligionModule.kt
package org.flossware.civilization.module

import org.flossware.civilization.model.*
import kotlin.random.Random

object ReligionModule {
    private const val SCHISM_THRESHOLD = 0.05
    private const val MINORITY_THRESHOLD = 0.2
    private const val UNITY_THRESHOLD = 0.6

    fun tick(current: ReligionState, tradeConnectivity: Double, random: Random): ModuleResult<ReligionState> {
        val events = mutableListOf<Event>()
        var newShares = spreadReligions(current.religionShares, current.spreadRate, tradeConnectivity, random)
        var unity = newShares.values.maxOrNull() ?: 0.0
        var stabilityBonus = unity * 0.2

        val schismOccurred = checkForSchism(newShares, unity, random)
        if (schismOccurred) {
            events.add(Event(0, "", EventType.RELIGIOUS_SCHISM, EventSeverity.MAJOR,
                "Religious schism splits the dominant faith!", null))
            newShares = splitLargestReligion(newShares, random)
            unity = newShares.values.maxOrNull() ?: 0.0
            stabilityBonus = unity * 0.2
        }

        val newState = ReligionState(newShares, unity, stabilityBonus, current.spreadRate)
        return ModuleResult(newState, events)
    }

    private fun spreadReligions(currentShares: Map<String, Double>, spreadRate: Double, tradeConnectivity: Double, random: Random): Map<String, Double> {
        val newShares = currentShares.toMutableMap()
        val effectiveSpreadRate = spreadRate * (1.0 + tradeConnectivity * 0.5)
        for ((religion, currentShare) in currentShares) {
            val change = (random.nextDouble() * 2.0 - 1.0) * effectiveSpreadRate
            newShares[religion] = (currentShare + change).coerceIn(0.0, 1.0)
        }
        return normalizeShares(newShares)
    }

    private fun checkForSchism(shares: Map<String, Double>, unity: Double, random: Random): Boolean {
        val hasSignificantMinority = shares.values.any { it > MINORITY_THRESHOLD && it < unity }
        return hasSignificantMinority && unity < UNITY_THRESHOLD && random.nextDouble() < SCHISM_THRESHOLD
    }

    private fun splitLargestReligion(shares: Map<String, Double>, random: Random): Map<String, Double> {
        val newShares = shares.toMutableMap()
        val largestReligion = shares.maxByOrNull { it.value }?.key ?: return shares
        val originalShare = shares[largestReligion] ?: return shares
        val splitRatio = 0.6 + (random.nextDouble() * 0.2 - 0.1)
        newShares[largestReligion] = originalShare * splitRatio
        newShares["$largestReligion (Reformed)"] = originalShare * (1.0 - splitRatio)
        return normalizeShares(newShares)
    }

    private fun normalizeShares(shares: Map<String, Double>): Map<String, Double> {
        val total = shares.values.sum()
        if (total <= 0) return shares
        return shares.mapValues { it.value / total }
    }
}
